server/client.go

```diff
@@ -452,6 +456,26 @@ func (c *client) processConnect(arg []byte) error {
 			c.authViolation()
 			return ErrAuthorization
 		}
+
+		// Check if opt-in tracing is enabled globally.
+		srv.mu.Lock()
+		globalOptInTrace := srv.opts.OptInTrace
+		srv.mu.Unlock()
+
+		// Check if opt-in tracing is enabled for this client.
+		c.mu.Lock()
+		clientOptInTrace := c.perms != nil && c.perms.trace
```

Comment:
This has to happen after the call to `checkAuth` I think because otherwise `c.perms` is nil.